lib/publiccloud/provider.pm

```diff
@@ -434,7 +501,7 @@ This method is called called after each test on failure or success.
 =cut
 sub cleanup {
     my ($self) = @_;
-    $self->terraform_destroy() if ($self->terraform_applied);
+    $self->terraform_destroy() if (($self->terraform_applied) || get_var('PUBLIC_CLOUD_ALWAYS_CLEANUP'));
```

Comment:
What was the reason for `PUBLIC_CLOUD_ALWAYS_CLEANUP` again?

Reply:
Partially answered above, but in that case if deployment fails (because of SAP installation issue, iSCSI server issue, or sporadic issues - mainly on Azure!) the `terraform_applied` state is still set to `0` and thus destroy is not done. And we don't want this for SAP.

Reply:
Does this mean, terraform_applied failed, but there are still some resources created?
So maybe we need to call terraform_destroy() just always if someone called terraform apply! Regardless of the return value.

Reply:
>Does this mean, terraform_applied failed, but there are still some resources created?

Yes, this could happen in our case: if VM deployment is OK but HANA installation fails for example.

>What happen if we call terraform destroy on a failed apply? Do you have a output of such call?

I don't have a copy no, but destroy can be called without issue most of the time. We can yes call this in any cases, I was just not sure if it was a good idea for your cases or not.

Reply:
Lets do it and avoid that variable. @asmorodskyi WDYT?

Reply:
yeah I would say that calling destroy unconditionally is better idea than introducing new variable